### DISCUSSION TITLE
https://starter.obytes.com/getting-started/environment-vars-config/

### DISCUSSION THREAD

**Discussion by giscus** [obytes/react-native-template-obytes#263] [Docs Comments]
# https://starter.obytes.com/getting-started/environment-vars-config/

Environment variables and configuration with validation and type-checking.

https://starter.obytes.com/getting-started/environment-vars-config/

**Comment by chanphiromsok** [comment]
Env.NAME is undefined when use to build android

**Reply by yjose** [reply]
Did you make sure to run the pre-build before building the project?

**Reply by chanphiromsok** [reply]
Hello it is my fault I add run node ./env.js to postinstall script

**Reply by chanphiromsok** [reply]
Hi again when I used two laptop when I use my 16Pro intel it can not run ./gradlew clean && ./gradlew assembleDebug whereas M1 ./gradlew clean && ./gradlew assembleDebug is ok in android folder did I miss something ?

**Reply by chanphiromsok** [reply]
Env is has value but when build android it error on 16ProIntel

**Reply by yjose** [reply]
yeah, you should run `pnpm run android` instead to make sure env vars are loading correctly

**Reply by chanphiromsok** [reply]
is that the only way ?

**Reply by yjose** [reply]
Yeah correct

**Reply by chanphiromsok** [reply]
now I write bashscript instead

**Reply by yjose** [reply]
I'm a bit curious about why you need to write a bash script instead of running `pnpm run android`🧐

**Reply by chanphiromsok** [reply]
I used to get error in android when running yarn android --variant=release in expo sdk49 that why I write bashscript

**Reply by rbshh** [reply]
While not the primary cause, it's worth checking if your .gitignore file unintentionally excludes the environment configuration files containing "Env.NAME". This can happen if you accidentally add patterns like .env or specific directory paths containing your environment files.

If this is the case, remove the relevant patterns from your .gitignore file to allow these files to be included during the build process. However, exercise caution and avoid committing sensitive information like API keys or passwords to your version control system.

**Comment by farrukhabbas** [comment]
I defined a number based env variable in the env.js, and assigned house: process.env.HOURS to the clientEnv object, then the pnpm ios fails with type error - cannot assign string to number, then when i convert the number like Number(process.env.HOURS) then the pre-commit fails with error cannot assign number to string, what can i do here? how to define a number variable ?

**Reply by yjose** [reply]
`dotenv` always reads env vars as strings, which is the normal behavior.  the solution is exactly what you did  but you need to  some fixes on typing, check  #296

**Comment by yjose** [comment]
For people trying to replicate the same approach in their own projects, a common issue we encounter

`Unable to resolve module path` means you are trying to import the path Node.js module into your app, which will not work because the runtime is different. This issue arises because you are importing the ./env.js file into a file located in the ./src folder, which Metro is attempting to bundle.

To resolve this, ensure you import your environment variables from @env, which should be an [alias](https://github.com/obytes/react-native-template-obytes/blob/a46693e6ddad22bd42e8c03fec623120d9d376d4/babel.config.js#L15) for ./src/core/env.js, instead of ./env.js.

**Comment by vlanemcev** [comment]
Hi, what do you think about new Expo Env handling model / approach? 
It seems to be more convenient to use thanks to Expo's built-in capabilities.
Interesting to see your opinion, thank you!

https://expo.dev/blog/environment-variables

**Reply by yjose** [reply]
@vlanemcev thank you for bringing this up for discussion.

When Expo introduced automatic loading of environment variables, I tried to adapt the approach. Unfortunately, it did not work because they primarily support `development` and `production` only. As a we are uning `staging` too, I disabled their feature using `EXPO_NO_DOTENV`. 

But now, as they introduce `preview`, which will be aligned with `staging`, I think we can adapt our approach to make it work smoothly with the new Expo approach. This should make it better and simpler for people to follow.

To be honest, I still haven't started working on it. I have some ideas, but feel free to share your thoughts on how we can adapt this to the new expo approach.

**Reply by waptik** [reply]
any updates on the ideas you have? also, where do you suggest we add api keys? or they should be used directly with `process.env`?
